core/src/main/scala/cats/evidence/Is.scala

```diff
@@ -0,0 +1,67 @@
+package cats.evidence
+
+import cats.Id
+
+/**
+ * A value of `A Is B` is proof that the types `A` and `B` are the same. More
+ * powerfully, it asserts that they have the same meaning in all type
+ * contexts. This can be a more powerful assertion than `A =:= B` and is more
+ * easily used in manipulation of types while avoiding (potentially
+ * erroneous) coercions.
+ *
+ * `A Is B` is also known as Leibniz equality.
+ */
+abstract class Is[A, B] {
+  def subst[F[_]](fa: F[A]): F[B]
+
+  @inline final def andThen[C](next: B Is C): A Is C =
+    next.subst[A Is ?](this)
+
+  @inline final def compose[C](prev: C Is A): C Is B =
+    prev andThen this
+
+  @inline final def from: B Is A =
```

Comment:
I'm not a huge fan of `from`, but I was taking it from some other sources. Calling it `reverse` feels alright since underlying `A Is B` is an actual process from `A` to `B`, but I sort of want to wipe that notion away. It's incidental instead of important. So, in order of my preference I like:

* `flip`
* `mirror`
* `reverse`
* `from`

But any of the above are acceptable to me.